### PR TITLE
unroll to unitary gate if unitary is in basis

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
When the Unroller pass encountered a gate which did not exist in the basis and had no definition, it would raise an error. This PR makes it possible to allow these types of gates if they have a unitary matrix representation and `unitary` is in the basis, such as for a simulator.

### CODE REVIEW

File: test/python/transpiler/test_unroller.py [hunk 1]
```diff
@@ -353,3 +356,38 @@ def test_unrolling_parameterized_composite_gates(self):
...
+        out_dag = pass_.run(dag)
+        self.assertEqual(len(out_dag.gate_nodes()), 1)
+        dag_op = out_dag.gate_nodes()[0].op
+        # assume decomposition of 3-qubit unitary is not in define for UnitaryGate
+        self.assertEqual(dag_op._definition, None)
+        self.assertTrue(numpy.array_equal(dag_op.to_matrix(), matrix))
+        self.assertEqual(dag_op.name, 'unitary')
+
+        pass_ = Unroller(['u1', 'u2', 'u3'])
+        self.assertRaisesRegex(NotImplementedError, 'Not able to generate a subcircuit*',
```
Comment (kdk) [inline test/python/transpiler/test_unroller.py]:
```suggestion
        self.assertRaisesRegex(NotImplementedError, 'Not able to generate a subcircuit',
```

File: qiskit/transpiler/passes/unroller.py [hunk 2]
```diff
@@ -68,9 +70,13 @@ def run(self, dag):
                 raise QiskitError('Error decomposing node {}: {}'.format(node.name, err))
 
             if not rule:
-                raise QiskitError("Cannot unroll the circuit to the given basis, %s. "
-                                  "No rule to expand instruction %s." %
-                                  (str(self.basis), node.op.name))
+                if 'unitary' in self.basis:
+                    rule = [(UnitaryGate(Operator(node.op), label=node.name),
```
Comment (kdk) [inline qiskit/transpiler/passes/unroller.py]: Is `UnitaryGate(node.op)` sufficient?

File: test/python/transpiler/test_unroller.py [hunk 3]
```diff
@@ -353,3 +356,38 @@ def test_unrolling_parameterized_composite_gates(self):
...
         self.assertEqual(circuit_to_dag(expected), out_dag)
+
+    def test_unroll_operator(self):
+        """test unrolling of gate without definition but a unitary
+        representation"""
+        sigmax = numpy.array([[0, 1], [1, 0]])
+        sigmay = numpy.array([[0, -1j], [1j, 0]])
+        matrix = numpy.kron(sigmay, numpy.kron(sigmax, sigmay))
+
+        class MyGate(Gate):
```
Comment (ajavadia) [inline test/python/transpiler/test_unroller.py]: maybe this can be called XYGate

Comment (1ucian0) [inline test/python/transpiler/test_unroller.py]:
```suggestion
        class XYGate(Gate):
```

File: qiskit/transpiler/passes/unroller.py (hunk 2 above)
Comment (ewinston) [reply qiskit/transpiler/passes/unroller.py]: The init of `Operator` handles a few more types of inputs than `UnitaryGate` so I used that to be safe.

**Comment by ewinston** [comment]
Since this is somewhat of an unusual use case this can probably on hold at this time.

**Pull request closed without merging**